src/Container/Docker/Manifest.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Container.Docker.Manifest (
+  ManifestJson (..),
+  ManifestJsonImageEntry (..),
+  decodeManifestJson,
+  getLayerPaths,
+  manifestFilename,
+) where
+
+import Control.DeepSeq (NFData)
+import Data.Aeson (
+  FromJSON,
+  ToJSON,
+  defaultOptions,
+  eitherDecode,
+  genericToEncoding,
+  parseJSON,
+  toEncoding,
+  withObject,
+  (.!=),
+  (.:),
+ )
+import Data.ByteString.Lazy qualified as ByteStringLazy
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Text (Text)
+import GHC.Generics (Generic)
+
+manifestFilename :: Text
+manifestFilename = "manifest.json"
+
+-- | The manifest of container image for the top-level image, and optionally for parent images
+-- that this image was derived from. It consists of an array of metadata entries.
+--
+-- Reference:
+-- * https://github.com/moby/moby/blob/master/image/spec/v1.2.md
+-- * https://github.com/moby/moby/blob/master/image/spec/v1.1.md
+newtype ManifestJson = ManifestJson (NonEmpty.NonEmpty ManifestJsonImageEntry)
+  deriving (Show, Eq, Ord, Generic)
+  deriving anyclass (NFData) -- explicitly use DeriveAnyClass strategy to avoid deriving-defaults warning.
+
+instance FromJSON ManifestJson
+instance ToJSON ManifestJson where
+  toEncoding = genericToEncoding defaultOptions
+
+data ManifestJsonImageEntry = ManifestJsonImageEntry
+  { config :: Text -- references another file in the tar which includes the image JSON for this image.
+  , repoTags :: [Text] -- references pointing to this image.
+  , layers :: NonEmpty.NonEmpty FilePath -- points to the filesystem changeset tars
+  }
+  deriving (Show, Eq, Ord, Generic, NFData)
+
+instance ToJSON ManifestJsonImageEntry where
+  toEncoding = genericToEncoding defaultOptions
+
+instance FromJSON ManifestJsonImageEntry where
+  parseJSON = withObject "" $ \o ->
+    ManifestJsonImageEntry
+      <$> o .: "Config"
+      <*> o .: "RepoTags" .!= []
```

Comment:
Are there scenarios where we would see no repo tags and pass through an empty `[]`? If so can you also add a test case for this? I don't see it represented in the spec file.

Reply:
I don't expect to, but I chose to stay on side of caution - I've removed the default assoc now